beacon_node/eth1/src/deposit_cache.rs

```diff
@@ -203,6 +226,55 @@ impl DepositCache {
             Ok((tree.root(), deposits))
         }
     }
+
+    /// Gets the deposit count at block height = block_number.
+    ///
+    /// Fetches the `DepositLog` that was emitted at or just before `block_number`
+    /// and returns the deposit count as `index + 1`.
+    ///
+    /// Returns `None` if block number queried is 0 or less than deposit_contract_deployed block.
+    pub fn get_deposit_count_from_cache(&self, block_number: u64) -> Option<u64> {
+        // Contract cannot be deployed in 0'th block
+        if block_number == 0 {
+            return None;
+        }
+        if block_number < self.deposit_contract_deploy_block {
+            return None;
+        }
+        // Return 0 if block_num queried is before first deposit
+        if let Some(first_deposit) = self.logs.first() {
+            if first_deposit.block_number > block_number {
+                return Some(0);
+            }
+        }
+        let index = self
+            .logs
+            .binary_search_by(|deposit| deposit.block_number.cmp(&block_number));
+        match index {
+            Ok(index) => return self.logs.get(index).map(|x| x.index + 1),
+            Err(prev) => {
+                return Some(
+                    self.logs
+                        .get(prev.saturating_sub(1))
+                        .map_or(0, |x| x.index + 1),
+                )
+            }
+        }
+    }
+
+    /// Gets the deposit root at block height = block_number.
+    ///
+    /// Fetches the `DepositLog` that was emitted at or just before `block_number`
+    /// and returns the deposit root at that state.
+    ///
+    /// Note: This method can be potentially optimized by not recreating the `DepositDataTree`
+    /// at every invocation and caching the tree upto the last added deposit.
+    pub fn get_deposit_root_from_cache(&self, block_number: u64) -> Option<Hash256> {
+        let index = self.get_deposit_count_from_cache(block_number)?;
+        let roots = self.roots.get(0..index as usize)?;
+        let tree = DepositDataTree::create(roots, index as usize, DEPOSIT_CONTRACT_TREE_DEPTH);
```

Comment:
As you mentioned, this is going to be fairly inefficient.

For our testnet, this means that if we sync a cache of `4,096` blocks we're going to have to find the root of a list of `16,384` for each of those blocks. I just benched a tree hash root of 16k hashes at 8ms, so we're looking at 32 secs of hashing just to fill the cache.

I think a fairly easy solution to this would be to attach a `DepositDataTree` and a `Vec<u64, Hash256>` to the deposit cache. Each time we import a deposit we add the deposit to the `DepositDataTree` and then push the new root into the `Vec`. In this scenario, we only build the deposit tree once and when we want to resolve a block number to a deposit root we can just binary search the `Vec`.

Thoughts?

Reply:
Yup, this sounds perfect :+1: